tracer/src/Datadog.Trace/Propagators/W3CTraceContextPropagator.cs

```diff
@@ -7,125 +7,438 @@
+                }
 
-                    Value = 00-4bf92f3577b34da6a3ce929d0e0e4736-00f067aa0ba902b7-01
-                    base16(version) = 00
-                    base16(trace-id) = 4bf92f3577b34da6a3ce929d0e0e4736
-                    base16(parent-id) = 00f067aa0ba902b7
-                    base16(trace-flags) = 01  // sampled
-                    Valid traceparent when caller didn’t sample this request:
+                if (!string.IsNullOrWhiteSpace(context.Origin))
+                {
+                    sb.Append("o:").Append(context.Origin).Append(';');
+                }
+
+                if (context.TraceContext?.Tags?.ToArray() is { Length: > 0 } tags)
+                {
+                    foreach (var tag in tags)
+                    {
+                        if (tag.Key.StartsWith("_dd.p.", StringComparison.Ordinal))
+                        {
+#if NETCOREAPP
+                            var key = tag.Key.AsSpan(start: 6);
+#else
+                            var key = tag.Key.Substring(startIndex: 6);
+#endif
 
-                    Value = 00-4bf92f3577b34da6a3ce929d0e0e4736-00f067aa0ba902b7-00
-                    base16(version) = 00
-                    base16(trace-id) = 4bf92f3577b34da6a3ce929d0e0e4736
-                    base16(parent-id) = 00f067aa0ba902b7
-                    base16(trace-flags) = 00  // not sampled
-                 */
+                            sb.Append("t.").Append(key).Append(':').Append(tag.Value).Append(';');
+                        }
+                    }
+                }
 
-                if (traceParent!.Length != 55 || traceParent[2] != '-' || traceParent[35] != '-' || traceParent[52] != '-')
+                if (sb.Length == 3)
                 {
-                    return false;
+                    // "dd=", we never appended anything
+                    return string.Empty;
                 }
 
-                char w3cSampled = traceParent[54];
-                if (traceParent[53] != '0' || (w3cSampled != '0' && w3cSampled != '1'))
+                // remove trailing ";"
+                if (sb[sb.Length - 1] == ';')
                 {
-                    return false;
+                    sb.Length--;
                 }
 
-                var samplingPriority = w3cSampled == '0' ? 0 : 1;
+                return StringBuilderCache.GetStringAndRelease(sb);
+            }
+            finally
+            {
+                StringBuilderCache.Release(sb);
+            }
+        }
+
+        internal static bool TryParseTraceParent(string header, out W3CTraceParent traceParent)
+        {
+            traceParent = default;
+
+            if (header == null!)
+            {
+                return false;
+            }
+
+            if (header.Length != 55 || header[2] != '-' || header[35] != '-' || header[52] != '-')
+            {
+                // validate format
+                return false;
+            }
+
+            if (header[0] != '0' || header[1] != '0')
```

Comment:
Note: There's a couple of implementation changes we need to make to the traceparent/tracestate parsing, like in this case we shoulda accept the header as long as it is two hex digits. Would you prefer to make those changes in this PR or a separate PR?

Reply:
I'll make them here, thanks.

Reply:
I've made most of the changes, but I still need to check that they are valid hex digits even if they are not `00`, and not allow `ff`. Leaving this open for tomorrow.

Reply:
Done.